### PR TITLE
enhance gameplay logic

### LINKED ISSUE #193: [Fix][Doctor Needed] Enhance and fix gameplay
### 1.
Normally, the game flow goes like this (or should go like this, ideally)

- Game has started
- Cards are dealt (hole cards) to each player... (this happens off chain, so you just have to imagine it, lol)
- After step 2, These players must undergo one betting round before three community cards are dealt.
- One betting round means every single player's bet are equal...
- This step has been checked using `game.highest_staker` value in the `after_play` function located in `systems/actions.cairo`, however this function to set the value of the highest staker is never done.
- When a player can be the highest staker is during a `raise` operation, or during an `all_in`, when the `all_in` amount is greater than the `game.current_bet`. You must know that the pots are dynamic. Anyways, fix this issue.
- And yes, write tests on the above to show it has been fixed. That means you have to use some functions in the `tests/setup.cairo` to setup your own gameplay environment in the test file you'll create.
### 2.
- In the same `after_play`, anytime the betting round has been concluded, and ready for restart, reset all player's and game's `current_bet` and the `game.highest_staker` value to 0. If there are other values you feel should be reset too, please do let me know.
- Write tests to show that this actually works.
- Why I'm emphasizing on the tests is that I don't wish to miss anything out

### 3.
- Fix bet spacing in the game_params..
- Anytime a player places a bet (except the all in), the bet amount must be in multiples of the `game_params.bet_spacing`. Say if `bet_spacing` is 20, then all bets should be in multiples of 20.
- Test what you've written too in any of the test files you created.

## Furthermore?

- The target branch is the v1 branch.
- Please when applying, state whether you’re new to the whole language and all. If you are, I could help you steadily (Though it’ll stress me out)… if you aren’t, better for me… I’ll leave everything in your hands 😌, and you won’t be permitted to paste errors in my dm.
- Your code must be **strongly typed** for better readability please 🙏🏻
- Please use all available traits possible at your disposal 
- Document your code please, in a professional manner. Please avoid using too many small comments as you should know, only comment on hard to understand logic, and maybe some explanations on how and why it works
- When done, @ your username in a comment on top of the function you touched. If there's one there already, append with a comma and a space. (`// @ExistingUsername, @YourUsername`, etc)... It's cooler to leave your signature behind, yes? 😋
- Each `//` should be accompanied by a space, before typing, lock in.
- Please avoid code repitition. There may be lines of codes that would be reused, please extract them into a separate function. Lock in your good PR quality please 🙏🏻
- May the force be with you ☪️✋🏻
- Ofcourse, if you have any questions (Which I bet you'd have, one billion per cent), you can branch [here](https://t.me/lemonade46) or [here](https://t.me/+tqBpITsr5mllZDQ0). Happy Hacks 😊💙

## Help?
"I made it here, and I will guide [this passage](https://github.com/Prometheus-A/The-Velvet-Ace/issues/193) all the way to the farthest ends of the earth. The stormiest seas will not stop me." _**--Pirate Gol D. Roger.**_

### PR DESCRIPTION
**Pull request by Kaylahray**
Closes #193 
### **Requirement**

* The `highest_staker` should be updated:

  * During a **raise** operation.
  * During an **all\_in** operation **only if** the all-in amount is **greater than** the game’s current bet.

### **Implementation**

* The `raise()` function in `actions.cairo` now **sets** `highest_staker`.
* The `all_in()` function includes logic to update `highest_staker` **only when** the all-in amount exceeds `game.current_bet`.

### **Tests**

* **`test_raise_sets_highest_staker`**
  Confirms that when a player successfully raises, they are correctly set as the `highest_staker` in the game state.

* **`test_all_in_sets_highest_staker_when_amount_greater_than_current_bet`**
  Verifies that a player who goes all-in with an amount **greater than** the current table bet becomes the new `highest_staker`.

* **`test_all_in_does_not_set_highest_staker_when_amount_less_than_current_bet`**
  Ensures that a player who goes all-in with an amount **less than** the current bet does **not** become the `highest_staker`.

### **Requirement**

* At the end of a betting round:

  * Reset `current_bet` for **all players** and the **game**.
  * Reset `game.highest_staker` to `None`.
* This must happen **automatically** when betting concludes and must be **tested**.

### **Implementation**
- Created two new internal functions, `is_betting_round_concluded` and `reset_betting_round_values`. This cleanly separates the logic for checking if a round has ended (all active, non-all-in players have equal bets) and the logic for resetting the game state (player bets, game bets, highest staker) for the next phase of the round.

* This function is called inside `after_play()` when `is_betting_round_concluded()` returns `true`.

### **Tests**

* **`test_betting_round_concludes_when_all_active_players_have_equal_bets`**
  Ensures that when all active players have equal bets and a non-betting action (e.g., `check`) occurs, the round concludes and all values (`current_bet`, `highest_staker`) are reset.

* **`test_betting_round_does_not_conclude_when_players_have_unequal_bets`**
  A control test confirming that if players have **unequal** bets, the betting round **does not** conclude, and the game state remains unchanged.

* **`test_betting_round_with_all_in_players_exempt_from_bet_matching`**
  Verifies that **all-in players** (who can’t raise further) **do not block** the round from ending. It confirms that the round ends when all other active players match the highest bet.

### **Requirement**

* Any **raise amount** must be a **multiple** of `game_params.bet_spacing`.
* This constraint must be **strictly enforced and tested**.

### **Implementation**

* The `raise()` function includes an assertion:

  ```cairo
  assert(raise_amount % game.params.bet_spacing == 0, 'Raise amount must be in multiples of bet_spacing');
  ```
This ensures invalid raises are rejected at runtime.

###  **Tests**

* **`test_bet_spacing_validation_in_raise`**
  Confirms that a player can successfully raise when their amount is a valid multiple of `game.params.bet_spacing`.

* **`test_bet_spacing_validation_fails_with_invalid_multiple`**
  Ensures that any raise not adhering to the bet spacing rule **correctly panics** and is not accepted.

image

### CODE REVIEW

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 1]
```diff
@@ -293,19 +293,23 @@ pub mod actions {
...
-                assert!(
-                    no_of_chips > game_pot * 2, "Raise amount should be > twice the small blind.",
-                );
+                assert!(no_of_chips > game_pot * 2, "Raise must be > 2x blind.");
             }
 
-            assert!(no_of_chips > 0, "Raise amount must be greater than zero.");
-            assert!(player.chips >= total_required, "You don't have enough chips to raise.");
+            assert!(no_of_chips > 0, "Raise must be > 0.");
+            assert!(player.chips >= total_required, "Insufficient chips to raise.");
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo:314]: I see good English. Welldone

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 2]
```diff
@@ -532,6 +566,88 @@ pub mod actions {
             self.world(@"poker")
         }
 
+        /// @Kaylahray Checks if betting round has concluded by verifying all active players have
+        /// equal bets
+        fn is_betting_round_concluded(
+            self: @ContractState, game_id: u64, world: @WorldStorage,
+        ) -> bool {
+            let game: Game = world.read_model(game_id);
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo]: You read the whole game model just to read the `players` value. Why not read the `players` "member" instead using `world.read_member()`

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 3]
```diff
@@ -776,74 +892,51 @@ pub mod actions {
             );
         }
 
-        /// @Reentrancy, @Birdmannn
         fn after_play(ref self: ContractState, caller: ContractAddress) {
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo]: please replace all the comments you deleted in this function because I don't consider this function as one that has been finalized. And replace the Birdmannn and that of Reentrancy's.

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 4]
```diff
@@ -776,74 +892,51 @@ pub mod actions {
...
+                assert(current_index_option.is_some(), 'Caller not in game');
+                let current_index = OptionTrait::unwrap(current_index_option);
+                let next_player_option = self
+                    .find_next_active_player(@game.players, current_index, @world);
+
+                if let Option::Some(next_player_addr) = next_player_option {
+                    game.next_player = Option::Some(next_player_addr);
+                } else {
+                    // If no next player, something is wrong, or round should end.
+                    // `is_betting_round_concluded` should have caught this.
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo]: How would I know? Did you test it? 😂

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 5]
```diff
@@ -532,6 +566,88 @@ pub mod actions {
...
+
+                if player.in_round {
+                    active_player_count += 1;
+                    if player.current_bet > highest_bet {
+                        highest_bet = player.current_bet;
+                    }
+                }
+                i += 1;
+            };
+
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo]: I don't think this would work out. I think the former implementation was just fine. `after_play` is called after each player has played, and checks Line588 for that particular player. You looping through all the players after each player has played just for the line588 is just gas.

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 6]
```diff
@@ -532,6 +566,88 @@ pub mod actions {
...
+                    if player.chips > 0 && player.current_bet != highest_bet {
+                        all_matched = false;
+                        break;
+                    }
+                }
+                j += 1;
+            };
+
+            all_matched
+        }
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo]: Please don't loop per player. If you can use an extra variable to keep track, please do.. or if the current player's bet matches the highest staker, and the next player is the highest staker

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 7]
```diff
@@ -532,6 +566,88 @@ pub mod actions {
...
+
+        /// @kaylahray Resets betting values when a round concludes
+        fn reset_betting_round_values(self: @ContractState, game_id: u64, ref world: WorldStorage) {
+            let game: Game = world.read_model(game_id);
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo]: Yup, and read the member, not the full Game.

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 8]
```diff
@@ -776,74 +892,51 @@ pub mod actions {
...
+
+                if let Option::Some(next_player_addr) = next_player_option {
+                    game.next_player = Option::Some(next_player_addr);
+                } else {
+                    // If no next player, something is wrong, or round should end.
+                    // `is_betting_round_concluded` should have caught this.
+                    game.showdown = true;
+                    game.next_player = Option::None;
+                }
             }
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo:869]: Please why did you delete the event? 👀

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 9]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
+
+        // [Setup State] - Set initial game state with current bet
+        let mut game: Game = world.read_model(1);
+        game.current_bet = 100;
+        world.write_model(@game);
+
+        let mut player_1: Player = world.read_model(PLAYER_1());
+        player_1.current_bet = 100; // Player has called the current bet
+        player_1.chips = 2000;
+        world.write_model(@player_1);
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: please extract this to a separate dynamic function as the code is repeated.
the function can return a tuple or a struct that contains the `context` of variables to be used later in the function if and where necessary

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 10]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
         // [Execute]
         // PLAYER_2 trying to play when it's PLAYER_1's turn
         set_contract_address(PLAYER_2());
         systems.actions.check();
     }
 
+    // ===== TESTS FOR HIGHEST STAKER AND BETTING ROUND LOGIC =====
+
+    #[test]
+    fn test_raise_sets_highest_staker() {
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: All these can be collapsed into one function though. That said, I'm focused on the gameplay, and not some obvious tests that should pass. Though you can do one or two of these to be sure it actually sets. The aim was to check with edge cases (if possible) if players `raise` and `all_in` actually goes round when it does... so at some point you have to test if `deal_community_card()` will work when it is supposed to, and `game.community_dealing` returns true or false... based on your poker knowledge. If you have any questions/confusions about the gameplay, you should ask.

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 11]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
+        // [Assert]
+        let game: Game = world.read_model(1);
+        let player_1_updated: Player = world.read_model(PLAYER_1());
+
+        assert_eq!(game.current_bet, 300, "Game current bet should be 300");
+        assert_eq!(
+            game.highest_staker, Option::Some(PLAYER_1()), "Player 1 should be highest staker",
+        );
+        assert_eq!(player_1_updated.current_bet, 300, "Player 1 current bet should be 300");
+        assert_eq!(player_1_updated.chips, 1800, "Player 1 should have 1800 chips left");
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: please remove these and merge this function with the next for the benefit of doubt

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 12]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
+
+        // [Setup State]
+        let mut game: Game = world.read_model(1);
+        game.current_bet = 500;
+        world.write_model(@game);
+
+        let mut player_1: Player = world.read_model(PLAYER_1());
+        player_1.current_bet = 0; // Player hasn't bet yet
+        player_1.chips = 1000; // Player has 1000 chips
+        world.write_model(@player_1);
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: as normal, extract all to that dynamic function I spoke of. Please know that that function should take in the mock parameters

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 13]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
+        assert_eq!(
+            game_updated.highest_staker, Option::None, "Highest staker should be reset to None",
+        );
+        assert_eq!(player_1_updated.current_bet, 0, "Player 1 current bet should be reset to 0");
+        assert_eq!(player_2_updated.current_bet, 0, "Player 2 current bet should be reset to 0");
+        assert_eq!(player_3_updated.current_bet, 0, "Player 3 current bet should be reset to 0");
+    }
+
+    #[test]
+    fn test_betting_round_does_not_conclude_when_players_have_unequal_bets() {
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: remove this test. You can assert this just before the betting actually concludes in the last test.. you might need just to change some values of just one player

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 14]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
+        assert_eq!(game.current_bet, 1500, "Game current bet should remain 1500");
+        assert_eq!(
+            game.highest_staker, Option::Some(PLAYER_2()), "Player 2 should remain highest staker",
+        );
+        assert_eq!(player_1_updated.current_bet, 1000, "Player 1 current bet should be 1000");
+        assert_eq!(player_1_updated.chips, 0, "Player 1 should have 0 chips left");
+    }
+
+    #[test]
+    fn test_betting_round_concludes_when_all_active_players_have_equal_bets() {
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: please simulate the gameplay and don't overuse the mocking of values because dojo gives you access to the storage. And as normal, any mocking, use that function we talked about 🌚👀

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 15]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
+
+        assert_eq!(game.current_bet, 1000, "Game current bet should be 1000");
+        assert_eq!(
+            game.highest_staker, Option::Some(PLAYER_1()), "Player 1 should be highest staker",
+        );
+        assert_eq!(player_1_updated.current_bet, 1000, "Player 1 current bet should be 1000");
+        assert_eq!(player_1_updated.chips, 0, "Player 1 should have 0 chips left");
+    }
+
+    #[test]
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: Please feel free to remove this test

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 16]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
+        let mut params = game.params;
+        params.bet_spacing = 20; // Set bet spacing to 20
+        game.params = params;
+        world.write_model(@game);
+
+        let mut player_1: Player = world.read_model(PLAYER_1());
+        player_1.current_bet = 100;
+        player_1.chips = 2000;
+        world.write_model(@player_1);
+
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: as normal, that function we spoke about earlier

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 17]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
+        set_contract_address(player_1.id);
+
+        // [Execute] - Player raises by 40 (multiple of 20) - should succeed
+        systems.actions.raise(140); // Total bet becomes 240, raise amount is 140
+
+        // [Assert]
+        let player_1_updated: Player = world.read_model(PLAYER_1());
+        assert_eq!(
+            player_1_updated.current_bet, 240, "Raise should succeed with valid bet spacing",
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: no need for this, please remove this block

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 18]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
+        game.params = params;
+        world.write_model(@game);
+
+        let mut player_1: Player = world.read_model(PLAYER_1());
+        player_1.current_bet = 100;
+        player_1.chips = 2000;
+        world.write_model(@player_1);
+
+        set_contract_address(player_1.id);
+
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: Merge this test with the previous by:
1. Removing this test
2. in the previous test, call Line361 with the next player
3. Yes, it'll panic, so add the `should_panic` macro at the top of the test.
4. And yes, when panicking, please remove the 'ENTRYPOINT_FAILED' message from the macro's `expected` value... for all your panic tests please.

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 19]
```diff
@@ -91,12 +91,327 @@ mod tests {
...
+        let mut player_3: Player = world.read_model(PLAYER_3());
+        player_3.current_bet = 1000;
+        player_3.chips = 4000;
+        world.write_model(@player_3);
+
+        set_contract_address(player_2.id);
+
+        // [Execute] - Player 2 checks, which should trigger betting round conclusion check
+        // Since all non-all-in players have equal bets, this should conclude the betting round
+        systems.actions.check();
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo]: if player_2 checks, then it should panic, because player 2 is the `highest_staker`. Though I don't understand what this test actually tests... but if this test passes, then you should understand the game flow and rework the logic. But first you must understand the game flow. 

Help me explain what this test does, or you can remove this test totally.

**Review by Birdmannn — COMMENTED** [review]
Neat code, but too bloated and quite some overhead shenanigans. Anyways, it seems you don't understand the poker gameplay, and their edge case scenarios... You didn't even test the scenario where the betting round ended and the variables were reset for another round to start. Remember, a betting round is different from the game round. 

Anyways, we are simulating gameplay, and not just mocking variables to test models' fields that have obviously changed. If you are to simulate by mocking, then you must explain the simulation using comments, though I'd still prefer you to use the abi calls for simulation. 

Please check. Thank you very much @Kaylahray. Nice job.
Please pay attention to the requested changes to avoid another review overhead. Thanks.

**Comment by Kaylahray** [comment]
@Birdmannn I've implemented every change requested and returned your comment and fixed my grammar 👀🫡

File: poker-texas-hold-em/contract/src/systems/actions.cairo (hunk 3 above)
Comment (Kaylahray) [reply poker-texas-hold-em/contract/src/systems/actions.cairo]: Replaced Sensei

File: poker-texas-hold-em/contract/src/systems/actions.cairo (hunk 2 above)
Comment (Kaylahray) [reply poker-texas-hold-em/contract/src/systems/actions.cairo]: image

File: poker-texas-hold-em/contract/src/systems/actions.cairo (hunk 3 above)
Comment (Kaylahray) [reply poker-texas-hold-em/contract/src/systems/actions.cairo]: image

File: poker-texas-hold-em/contract/src/systems/actions.cairo (hunk 4 above)
Comment (Kaylahray) [reply poker-texas-hold-em/contract/src/systems/actions.cairo]: 👀🙇‍♀️

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 20]
```diff
@@ -661,6 +342,113 @@ mod tests {
...
+
+        // Check all player states immediately after the call
+        game = world.read_model(1);
+        let player_1: Player = world.read_model(PLAYER_1());
+        let player_2: Player = world.read_model(PLAYER_2());
+        let player_3: Player = world.read_model(PLAYER_3());
+
+        // At this point, all players should have current_bet = 40, game.current_bet = 40
+        // The betting round should be complete and reset should have happened
+        assert!(game.highest_staker.is_none(), "Highest staker not reset");
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo:462]: Then the reset is complete. Check the community_dealing. As per last review on the tests, and the issue description, community dealing must be tested to. After first first round of betting, `deal_community_card()` must go through three times. 


Add test: The fourth one must panic
Add test: After adding three community cards, the betting should resume
Add test: Try any of the betting functions when the betting round has ended. This should panic.


and this just means that the betting cycle you wrote above can be extracted into a separate function called `feign_betting_round()` or something, and can be reused in any of the three tests as above.

There are other edge cases tests, but I'll leave it for another issue.

File: poker-texas-hold-em/contract/src/systems/actions.cairo [hunk 21]
```diff
@@ -891,48 +806,118 @@ pub mod actions {
...
+            for player_addr in game.players.span() {
+                let p: Player = world.read_model(*player_addr);
+                if p.in_round {
+                    active_players += 1;
+                    if p.current_bet != *game.current_bet {
+                        all_equal_bets = false;
+                        break;
+                    }
+                }
+            };
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/systems/actions.cairo]: In the last review, I said you shouldn't loop per player, it's just brute force... eats gas. what if there are thirty players at the table?

By the way, didn't this logic over here work?
Screenshot 2025-08-06 at 11 15 56 AM
no?

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 22]
```diff
@@ -661,6 +342,113 @@ mod tests {
...
+        let raise_amount = game.params.small_blind * 4; // 40 is multiple of bet_spacing (20)
+        set_contract_address(PLAYER_1());
+        systems.actions.raise(raise_amount.into());
+
+        let updated_game: Game = world.read_model(1);
+        assert_eq!(updated_game.current_bet, raise_amount.into(), "Bet spacing success failed");
+    }
+    //   @kaylahray Testing all-in sets highest staker
+    #[test]
+    fn test_all_in_sets_highest_staker() {
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo:502]: As per the issue description, The aim was for all_in to work regardless of the bet_spacing. So write a new test that mocks the player's balance to not be a multiple of the bet. spacing and call the all_in. It should pass

**Review by Birdmannn — COMMENTED** [review]
@Kaylahray check

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo (hunk 20 above)
Comment (Kaylahray) [reply poker-texas-hold-em/contract/src/tests/test_actions.cairo:462]: image
image


Trying two betting functions 👇
image

File: poker-texas-hold-em/contract/src/systems/actions.cairo (hunk 21 above)
Comment (Kaylahray) [reply poker-texas-hold-em/contract/src/systems/actions.cairo]: image

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo (hunk 22 above)
Comment (Kaylahray) [reply poker-texas-hold-em/contract/src/tests/test_actions.cairo:502]: image

**Comment by Kaylahray** [comment]
image
  @Birdmannn

File: poker-texas-hold-em/contract/src/tests/test_actions.cairo [hunk 23]
```diff
@@ -518,83 +518,233 @@ pub mod tests {
...
+        );
+
+        // Now deal three community cards (the flop)
+        let card1 = Card { suit: 4, value: 1 }; // Hearts Ace
+        let card2 = Card { suit: 1, value: 13 }; // Spades King
+        let card3 = Card { suit: 3, value: 12 }; // Diamonds Queen
+
+        systems.actions.deal_community_card(card1, 1);
+        systems.actions.deal_community_card(card2, 1);
+        systems.actions.deal_community_card(card3, 1);
```
Comment (Birdmannn) [inline poker-texas-hold-em/contract/src/tests/test_actions.cairo:589]: Can you see that this block is repeated? Three good times...

**Review by Birdmannn — COMMENTED** [review]
Anyways, Thank you for your sound understanding of the poker game (and your willingness to learn) 💯⛩️
